server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -650,19 +652,21 @@ private static Tuple<Map<String, DataStream>, Map<String, DataStreamAlias>> getD
                         )
                     );
                 } else if (featureStateDataStreams.contains(requestedDataStream)) {
-                    dataStreams.put(requestedDataStream, dataStreamInSnapshot);
+                    allDataStreams.put(requestedDataStream, dataStreamInSnapshot);
+                    systemDataStreams.put(requestedDataStream, dataStreamInSnapshot);
                 } else {
                     logger.debug(
                         "omitting system data stream [{}] from snapshot restoration because its feature state was not requested",
                         requestedDataStream
                     );
                 }
             }
-            if (includeAliases) {
+            if (includeAliases || systemDataStreams.isEmpty() == false) {
                 dataStreamAliases = new HashMap<>();
                 final Map<String, DataStreamAlias> dataStreamAliasesInSnapshot = globalMetadata.getProject().dataStreamAliases();
+                Map<String, DataStream> dataStreamsWithAliases = includeAliases ? systemDataStreams : allDataStreams;
```

Comment:
Wouldn't this need to be inverted? So `includeAliases ? allDataStreams : systemDataStreams`?

Might also be worth adding a small comment to explain why we're doing that.